### PR TITLE
adds tagging support for events (closes #1907)

### LINKED ISSUE #1907: Events Service does not support Tagging
Terraform fails when attempting to create a CloudWatch Event rule, this has been determined to be related to the lack of a tagging implementation within Events.

### PR DESCRIPTION
**Pull request by bryanlalexander**
adds localstack.utils.TaggingService, implements TaggingService for CloudWatch Events

**Please refer to the contribution guidelines in the README when submitting PRs.**

### CODE REVIEW

File: localstack/utils/tagging.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    def tag_resource(self, arn, tags):
+        if arn not in self.tags:
+            self.tags[arn] = {}
+        for t in tags:
+            self.tags[arn][t['Key']] = t['Value']
+
+    def untag_resource(self, arn, tag_names):
+        for name in tag_names:
+            if name in self.tags[arn]:
```
Comment (whummer) [inline localstack/utils/tagging.py]: Better:
```
if name in self.tags.get(arn, {}):
    ...
```
(to secure against `KeyError`)

**Comment by bryanlalexander** [comment]
@whummer the following changes were made, some of which were not on your list, but we adding integration testing and fixed a couple bugs we found with the downloader.

- moved test_tagging.py to localstack/tests/unit
- fixed inconsistency in test_create_tag (map has no guaranteed key order)
- adds status code and size check to localstack.utils.common.download method
- adds exception logging for localstack.services.install.download_and_extract_with_retry
- fixed deprecated call to assertRegexMatches
- adds integration testing for ListTagsForResource, TagResource, and UntagResource
- adds constants import of APPLICATION_AMZ_JSON_1_1 to events_listener
- moves parsed_data to the foremost if statement in ProxyListenerEvents
- fixed bug in AWSEvents.UntagResource method in ProxyListenerEvents
- updates make_response in events_listener to do the json marshalling
- updates AWSEvents.PutEvents to use make_response in ProxyListenerEvents
- adds updated KeyError-safe check to TaggingService

File: localstack/utils/common.py [hunk 2]
```diff
@@ -607,8 +611,13 @@ def download(url, path, verify_ssl=True):
                     LOG.debug('Empty chunk %s (total %s) from %s' % (chunk, total, url))
             f.flush()
             os.fsync(f)
+        if os.path.getsize(path) == 0:
+            LOG.error('Zero bytes downloaded from %s, retrying' % url)
```
Comment (whummer) [inline localstack/utils/common.py]: Can we please reduce the log level to `LOG.info` or `LOG.warning`? We're generally only using `error` for fatal errors.

File: localstack/services/install.py [hunk 3]
```diff
@@ -238,8 +238,9 @@ def download_and_extract():
 
     try:
         download_and_extract()
-    except Exception:
+    except Exception as e:
         # try deleting and re-downloading the zip file
+        LOGGER.error('download_and_extract_with_retry caught exception: %s' % e)
```
Comment (whummer) [inline localstack/services/install.py]: Can we combine the two log messages into one (with log level `info` or `warning`)? Something like
```
LOGGER.info('Unable to extract file, re-downloading ZIP archive %s: %s' % (tmp_archive, e))
```

**Comment by bryanlalexander** [comment]
@whummer All set, those changes have been made, thanks for the thorough review.

**Comment by whummer** [comment]
Awesome, thanks again!

**Pull request merged**